File: 2° Periodo/Prog 2  JAVA C++/POO_TRABALHO_1/src/Consultorio.java
import java.util.ArrayList;

public class Consultorio {
    private ArrayList<Medico> medicos;
    private ArrayList<Paciente> pacientes;
    private ArrayList<Consulta> consultas;
    private int id;

    public Consultorio() {
        this.medicos = new ArrayList<Medico>();
        this.consultas = new ArrayList<Consulta>();
        this.pacientes = new ArrayList<Paciente>();
        id = 0;
    }

    public void cadastrarPaciente(Paciente p) {
        for (Paciente i : pacientes)
            if (i.getCpf().equals(p.getCpf())) {
                System.out.println("Paciente já Cadastrado");
                return;
            }
        pacientes.add(p);
        System.out.println("Paciente Cadastrado");
    }

    public void removerPaciente(String cpf) {
        for (Paciente i : pacientes)
            if (i.getCpf().equals(cpf)) {
                pacientes.remove(i);
                System.out.println("Paciente Removido");
                return;
            }
        System.out.println("Paciente não Encontrado");
    }

    public void cadastrarMedico(Medico m) {
        for (Medico i : medicos)
            if (i.getCrm() == m.getCrm()) {
                System.out.println("Medico já Cadastrado");
                return;
            }
        medicos.add(m);
        System.out.println("Medico Cadastrado");
    }

    public void removerMedico(int crm) {
        for (Medico i : medicos)
            if (i.getCrm() == crm) {
                medicos.remove(i);
                System.out.println("Medico Removido");
                return;
            }
        System.out.println("Medico não Encontrado");
    }

    public void verificarId() {
        if (id == 500)
            id = 0;
        else
            id++;
    }

    public void cadastrarConsulta(Consulta c) {

        for (Consulta i : consultas) {
            if (i.getData().equals(c.getData()) && i.getHora().equals(c.getHora())
                    && i.getCpfPaciente().equals(c.getCpfPaciente()) && i.getCrmMedico() == c.getCrmMedico()) {
                System.out.println("Consulta já Cadastrado");
                return;
            } else if (i.getData().equals(c.getData()) && i.getHora().equals(c.getHora())
                    && i.getCrmMedico() == c.getCrmMedico()) {
                System.out.println("Horario já utilizado pelo Medico");
                return;
            }
        }

        for (Medico i : medicos) {
            if (i.getCrm() == c.getCrmMedico()) {
                for (Paciente j : pacientes) {
                    if (j.getCpf().equals(c.getCpfPaciente())) {

                        c.setId(id);
                        verificarId();

                        consultas.add(c);
                        System.out.println("Consulta Cadastrado");
                        c.imprimir();
                        return;
                    }
                }
            }
        }

        System.out.println("Medico ou Pacinete não Cadastrado");
    }

    public void removerConsulta(int id, int crmMedico, String cpfPaciente) {

        for (Consulta i : consultas)
            if (i.getCrmMedico() == crmMedico && i.getId()==id && i.getCpfPaciente().equals(cpfPaciente)){
                consultas.remove(i);
                System.out.println("Consulta Removido");
                return;
            }
        System.out.println("Consulta não Encontrado");
    }

    public void imprimirListaPacientes() {
        if (pacientes.isEmpty()) {
            System.out.println();
            System.out.println("Nenhum Paciente Cadastrado\n");
            return;
        }

        for (Paciente i : pacientes) {
            i.imprimir();
            System.out.println("\n");
        }
    }

    public void imprimirListaMedicos() {
        if (medicos.isEmpty()) {
            System.out.println();
            System.out.println("Nenhum Medico Cadastrado");
            return;
        }
        for (Medico i : medicos) {
            i.imprimir();
            System.out.println("\n");
        }
    }

    public void imprimirListaConsultas() {
        if (consultas.isEmpty()) {
            System.out.println();
            System.out.println("Nenhum Consulta Cadastrada");
            return;
        }
        for (Consulta i : consultas) {
            i.imprimir();
            System.out.println("\n");
        }
    }

    public void imprimirConsulta(String cpfPaciente, int crmMedico, int id) {
        for (Consulta i : consultas) {
            if (i.getCpfPaciente().equals(cpfPaciente) && i.getCrmMedico() == crmMedico && i.getId()==id) {
                i.imprimir();
                return;
            }
        }
        System.out.println("Consulta não encontrado!\n");
    }
}

File: 2° Periodo/Prog 2  JAVA C++/LAB 10/src/Consultorio.cpp
#include "Consultorio.hpp"

Consultorio::Consultorio(){
    lista_consulta = new list<Consulta>();
    lista_medico = new list<Medico>();
    lista_paciente= new list<Paciente>();
    contPaciente = 0;
}

void Consultorio::cadastrarMedico(Medico m){
    lista_medico->push_back(m);
}

void Consultorio::removerMedico(int crm){
    list<Medico>::iterator aux;

    for(aux = lista_medico->begin(); aux != lista_medico->end(); aux++)
        if(aux->getCRM()==crm){
            aux->imprimir();
            lista_medico->erase(aux);
            return;
        }

    throw ExcecaoMedicoInexistente();
}

void Consultorio::cadastrarPaciente(Paciente p){
    lista_paciente->push_back(p);
    contPaciente++;
}

void Consultorio::removerPaciente(int cpf){
    list<Paciente>::iterator aux;

    for(aux = lista_paciente->begin(); aux != lista_paciente->end(); aux++)
        if(aux->getCpf()==cpf){
            aux->imprimir();
            lista_paciente->erase(aux);
            contPaciente--;
            return;
        }

    throw ExcecaoPacienteInexistente();
}

void Consultorio::cadastrarConsulta(Consulta co){
    lista_consulta->push_back(co);
}

void Consultorio::removerConsulta(int cpf, string data){
    list<Consulta>::iterator aux;

    for(aux = lista_consulta->begin();aux != lista_consulta->end(); aux ++)
        if(aux->getData()==data && aux->getPaciente().getCpf()==cpf){
            cout<<"\nNome do Paciente: "<<aux->getPaciente().getNome()
            <<" do CPF "<<aux->getPaciente().getCpf()<<endl;
             cout<<"Nome do Medico: "<<aux->getMedico().getNome()
            <<" do CRM "<<aux->getMedico().getCRM()<<endl;
            cout<<"Data: "<< aux->getData()<<endl;
            cout<<"Horas: "<< aux->getHora()<<endl;
            lista_consulta->erase(aux);
            return;
        }

    throw ExcecaoConsultaInexistente("Impossível remover Consulta. Paciente " + to_string(cpf) + " não tem consulta agendada para a data " + data);

}

Paciente Consultorio::buscarPaciente(int cpf) {
    list<Paciente>::iterator aux;

    for (aux = lista_paciente->begin(); aux != lista_paciente->end(); ++aux) {
        if (aux->getCpf() == cpf) {
            return *aux;
        }
    }

    throw ExcecaoPacienteInexistente("Paciente com CPF " + to_string(cpf) + " não encontrado");
}

Medico Consultorio::buscarMedico(int crm){
     list<Medico>::iterator aux;

     for(aux = lista_medico->begin(); aux != lista_medico->end(); aux++){
        if (aux->getCRM()==crm){
            return *aux;
        }
     }

    throw ExcecaoMedicoInexistente("Medico do CRM " + to_string(crm) + " não encontrado");

}

Consulta Consultorio::buscarConsulta(int cpf,int crm, string data){
    list<Consulta>::iterator aux;

    for(aux = lista_consulta->begin(); aux != lista_consulta->end(); aux++){
        if(aux->getMedico().getCRM() == crm && aux->getPaciente().getCpf()==cpf && aux->getData()==data){
            return *aux;
        }
    }

throw ExcecaoConsultaInexistente("Consulta não encontrado");

}

void Consultorio::imprimirConsultas(){
    if(lista_consulta->empty())
        throw ExcecaoConsultaInexistente("Lista de Consultas esta vazia");

    list<Consulta>::iterator aux;

    for(aux = lista_consulta->begin();aux != lista_consulta->end(); aux ++){
        cout<<"\nNome do Paciente: "<<aux->getPaciente().getNome()
            <<" do CPF "<<aux->getPaciente().getCpf()<<endl;
        cout<<"Nome do Medico: "<<aux->getMedico().getNome()
            <<" do CRM "<<aux->getMedico().getCRM()<<endl;
        cout<<"Data: "<< aux->getData()<<endl;
        cout<<"Horas: "<< aux->getHora()<<endl;
        cout<<"\n---------------\n";
    }
}

void Consultorio::imprimirListaDeMedicos(){
    if(lista_medico->empty())
        throw ExcecaoMedicoInexistente("Lista de Medicos esta vazia");

    list<Medico>::iterator aux;

    for(aux = lista_medico->begin();aux != lista_medico->end(); aux ++){
        aux->imprimir();
        cout<<"\n---------------\n";
    }
}

void Consultorio::imprimirListaDeTodosOsPaciente(){
    if(lista_paciente->empty())
        throw ExcecaoPacienteInexistente("Lista de Pacientes esta vazia");

    list<Paciente>::iterator aux;

    for(aux = lista_paciente->begin();aux != lista_paciente->end(); aux ++){
        aux->imprimir();
        cout<<"\n---------------\n";
    }
}